Add Global Time selection palette to Clock activity

nithinshiju-dev opened the pull request:
This PR adds a new “Global Time” feature to the Clock activity.
Learners can choose a city/timezone from a Sugar palette and the clock updates to display the current time for the selected location.

A new Global Time toolbar button opens the palette, which lists available timezones.
When “Local Time” is selected, the clock behaves as before using the device time.
For other selections, the time is computed using standard IANA timezone support.

The selected timezone is displayed next to the time when it is non-local, all labels are localized, and the user’s choice is saved to the Journal so it is restored when reopening the activity.

Fixes #1825 

<img width="1920" height="1200" alt="Screenshot 2025-12-05 185621" src="https://github.com/user-attachments/assets/a7f2b269-7711-49a1-86ff-23b1c895befa" />

nithinshiju-dev committed:
Add Global Time picker with timezone cities to Clock activity

nithinshiju-dev committed:
Add global time palette with localization and persistence

nithinshiju-dev committed:
Fix Global Time palette layout and localization

nithinshiju-dev commented:
Hi @llaske,

Thanks for the feedback! I’ve addressed all the points:

* The PR is now based on the `dev` branch.
* The combobox has been replaced with a Sugar-style palette list.
* All texts required for the Global Time feature have been added to `activities/Clock.activity/locales/en.json`. Please let me know if you’d like me to add the same entries to the other locale files as well.
* When a non-local timezone is selected, its name is displayed on the screen.
* The selected timezone is saved in the Journal and restored when the activity is reopened.

Based on your feedback, I can also implement the locales for the remaining languages if needed.

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -1,5 +1,9 @@
-define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiobuttonsgroup","mustache","moment-with-locales.min","l10n", "tutorial"], function (activity,env,radioButtonsGroup,mustache,moment, l10n, tutorial) {
-
+define(
```
Hey Kindly fix this indentation of line

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -234,17 +296,99 @@ define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiob
                 show_mins = document.getElementById("show-mins").classList.contains("active");
                 that.drawBackground();
             });
+
```
This too plz

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -234,17 +296,99 @@ define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiob
+            var date = getCurrentDate();
             var hours = date.getHours();
-            if (hours<12){
+            if (hours < 12) {
```
here as well

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -234,17 +296,99 @@ define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiob
 
-        function setTranslatedStrings() {
+
+          function setTranslatedStrings() {
```
Here too

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -255,8 +399,14 @@ define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiob
+            }
         }
 
+
```
here as well

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -381,26 +531,35 @@ define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiob
         // Update text and hand angles using the current time.
-        Clock.prototype.update = function () {
-            var date = new Date();
+          Clock.prototype.update = function () {
```
a extra tab is added

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -381,26 +531,35 @@ define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiob
+                    displayHours = hours;
+                } else {
+                    displayHours = hours - 12;
+                }
```
Indentation problem

SnehalSrivastava27 reviewed (commented):
Hey there are some indentation issue please fix it

nithinshiju-dev committed:
Fix indentation and formatting in Clock activity

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -1,5 +1,16 @@
-define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiobuttonsgroup","mustache","moment-with-locales.min","l10n", "tutorial"], function (activity,env,radioButtonsGroup,mustache,moment, l10n, tutorial) {
-
+define([
```
<img width="1109" height="455" alt="Image" src="https://github.com/user-attachments/assets/9e342868-b82f-49a3-a9e0-c936784a8002" />

See this @nithinshiju-dev

nithinshiju-dev replied:
Hi @SnehalSrivastava27 — I saw the screenshot but I’m not sure what exact change you want me to apply. Could you please clarify what needs fixing?

nithinshiju-dev replied:
and in this line: if (hours < 12) {  
am i supposed to remove the space between '<' ?

SnehalSrivastava27 replied:
Bro there is a pattern followed in every file 
the define always remain in the single line at the top of line but you have extended it to 10 lines 
So please edit that
Basically, the motto is not to touch or hamper the old indentation

nithinshiju-dev committed:
Restore single-line define formatting

nithinshiju-dev commented:
Hi @SnehalSrivastava27, is it fine now?

SnehalSrivastava27 commented on activities/Clock.activity/js/activity.js:
```diff
@@ -499,7 +652,7 @@ define(["sugar-web/activity/activity","sugar-web/env","sugar-web/graphics/radiob
                 }
               } else {
-                if (hours<12){
+                if (hours < 12) {
```
Remove extra space
Not a very big issue though but still

SnehalSrivastava27 commented:
> Hi @SnehalSrivastava27, is it fine now?

Yes, Thankyou

nithinshiju-dev commented:
@SnehalSrivastava27  Thanks for confirming—should I mention llaske for merging, or is there anything else I should do?

SnehalSrivastava27 commented:
> @SnehalSrivastava27 Thanks for confirming—should I mention llaske for merging, or is there anything else I should do?

No need to mention He will see the PR and review it

llaske commented:
Good job.

Few remarks:
- The city should appear even if digital hour is not displayed
- Will be nice to add UTC shift (UTC+/- X) with each city
- Should be better to sort cities by UTC
- Font size in the palette is small, but of course there is lot of items. May be we could reduce items by grouping on the same line cities with the same UTC?
- Will be better to add a vertical bar before the global time icon
- Add new locales strings in all languages files to avoid displaying the id. If you've got a doubt on localization keep the english version
- Use CSS media to hide the new icon when screen is resized to keep the Stop button visible as long as possible

https://github.com/user-attachments/assets/1927fdab-b9bc-4101-aa4e-b62c594cce13

nithinshiju-dev committed:
Sort grouped world time cities by UTC

nithinshiju-dev committed:
Add Global Time localization keys to all locales

nithinshiju-dev commented:
Hi @llaske, I’ve addressed all your remarks. Is there anything else you’d like me to change?
<img width="1920" height="1200" alt="Screenshot 2025-12-08 211539" src="https://github.com/user-attachments/assets/7f93faf4-d6ea-4e55-8579-54dd3f16eb2d" />

llaske commented:
Nice.
My feedback:
- The city should not be displayed if local time is choose (so it will appear like before if no time is clicked)
- The city selection in the palette is complex. I suggest to select by line (like for local time). So only the first city of the line will be display on the clock but we could add a fun trick: if the same line is selected, the second city will be display. Then the third (if there is one), then the first.
- The CSS media resize handling could be improved

https://github.com/user-attachments/assets/c46be8ce-ad46-44ff-b600-18c01eeb0969

nithinshiju-dev committed:
Improve Global Time palette resizing and city selection

nithinshiju-dev commented:
@llaske  All feedback has been addressed and pushed to the dev branch. Let me know if any further changes are needed.
<img width="1920" height="1200" alt="Screenshot 2025-12-10 174513" src="https://github.com/user-attachments/assets/b53543e8-7a8a-4d97-ab36-e413d2031eba" />

llaske commented:
- In local time mode the clock should use the whole height - here you keep the space for an invisible text
- Will be better to have the current line selected when opening the palette
- The media query CSS to keep the Stop button displayed could be improved

nithinshiju-dev committed:
Improve palette selection and toolbar responsiveness

nithinshiju-dev commented:
@llaske Sorry, in my previous push I forgot to include the resizing changes. They are now committed and pushed along with the other fixes.

https://github.com/user-attachments/assets/789228e4-badf-4058-9db4-2872e3b89a58

llaske commented:
There is a side effect on digital time display. It is not displayed without date.

https://github.com/user-attachments/assets/d939f8cb-8483-4f14-acd6-0868037c3127

nithinshiju-dev committed:
Fix digital time disappearing and layout issues in local timezone

nithinshiju-dev commented:
Sorry @llaske for needing multiple PR commits to fix this, I was still getting familiar
with the code structure, but I understand it much better now and will improve going forward.
The latest push includes the fix for the digital time issue in local mode.

https://github.com/user-attachments/assets/0dfac50a-dbed-4ea5-aae5-d0845c51572c

llaske commented:
> Sorry @llaske for needing multiple PR commits to fix this, I was still getting familiar with the code structure, but I understand it much better now and will improve going forward. 

Not a problem. We'll spent all time need to obtain the best result.

New issue: The palette selection is not set when the activity is reopen

<img width="1106" height="672" alt="image" src="https://github.com/user-attachments/assets/b0c34dae-1d42-4806-8540-28b68209099a" />

nithinshiju-dev committed:
Fix Global Time palette state

nithinshiju-dev commented:
@llaske This fixes an issue where the Global Time palette did not reflect the previously selected timezone when reopening the activity.

nithinshiju-dev commented:
Hi @llaske ,
I noticed a difference in behavior,
when launching an activity from Home, the previous activity state is restored, but when launching the same activity from the Menu, it starts fresh.
Is this the expected behavior, or should the Menu also restore the previous state?

https://github.com/user-attachments/assets/ca10cba9-d8cc-4d0c-907a-830119b0bca1

llaske commented:
> I noticed a difference in behavior, when launching an activity from Home, the previous activity state is restored, but when launching the same activity from the Menu, it starts fresh. Is this the expected behavior, or should the Menu also restore the previous state?

Yes it's the expected behavior. From the list view a new instance is always created. From the home view it start the latest instance or allow you to choose the instance to start using the popup menu.

Another issue: you should add a step in the tutorial for the new icon. It look like you've added the string but miss to add a new step to display it:

<img width="529" height="346" alt="image" src="https://github.com/user-attachments/assets/59a1e872-e92f-45bb-8f26-594777f8ace2" />

nithinshiju-dev commented:
Thanks for the clarification Sir. I’ll work on updating the tutorial.

nithinshiju-dev committed:
Add Global Time tutorial step and translations

nithinshiju-dev commented:
Hi @llaske! I’ve added a tutorial step for the Global Time icon and updated the text across all locales. The translations were generated using Google Translate.
<img width="1920" height="1200" alt="Screenshot 2025-12-14 155627" src="https://github.com/user-attachments/assets/4e97cfd6-548b-4123-83c0-4f3886f0770a" />
<img width="1920" height="1200" alt="Screenshot 2025-12-14 160303" src="https://github.com/user-attachments/assets/bf6683f9-2ac3-4849-8c40-458d0d8d44d8" />

llaske commented:
Nice. Thanks @nithinshiju-dev for your contribution.
